flocker/control/_protocol.py

```diff
@@ -0,0 +1,166 @@
+
+
+class VersionCommand(Command):
+    """
+    Return configuration protocol version of the control service.
+
+    Semantic versioning: Major version changes implies incompatibility.
+    """
+    arguments = []
+    response = [('major', Integer())]
+
+
+class ClusterStatusCommand(Command):
+    """
+    Used by the control service to inform a convergence agent of the
+    latest cluster state and desired configuration.
+
+    Having both as a single command simplifies the decision making process
+    in the convergence agent during startup.
+    """
+    arguments = [('configuration', DeploymentArgument()),
+                 ('state', DeploymentArgument())]
+    response = []
+
+
+class NodeStateCommand(Command):
+    """
+    Used by a convergence agent to update the control service about the
+    status of a particular node.
+    """
+    arguments = [('hostname', String()),
+                 ('node_state', NodeStateArgument())]
+    response = []
+
+
+class ControlServiceLocator(CommandLocator):
+    """
+    Control service side of the protocol.
+    """
+    def __init__(self, cluster_state):
+        """
+        :param ClusterStateService cluster_state: Object that records known
+            cluster state.
+        """
+        self.cluster_state = cluster_state
+
+    @VersionCommand.responder
+    def version(self):
+        return {"major": 1}
+
+    @NodeStateCommand.responder
+    def node_changed(self, hostname, node_state):
+        self.cluster_state.update_node_state(hostname, node_state)
+        return {}
+
+
+class IConvergenceAgent(Interface):
+    """
+    The agent that will receive notifications from control service.
+
+    This is a little sketchy; it will be solidified in
+    https://clusterhq.atlassian.net/browse/FLOC-1255
+    """
+    def connected():
+        """
+        The client has connected to the control service.
+        """
+
+    def disconnected():
+        """
+        The client has disconnected from the control service.
+        """
+
+    def cluster_updated(configuration, cluster_state):
+        """
+        The cluster's desired configuration or actual state have changed.
+
+        :param Deployment configuration: The desired configuration for the
+            cluster.
+
+        :param Deployment cluster_state: The current state of the
+            cluster. Mostly useful for what it tells the agent about
+            non-local state, since the agent's knowledge of local state is
+            canonical.
+        """
+
+
+class AgentClient(AMP):
+    """
+    Convergence agent side of the protocol.
+    """
+    def __init__(self, agent):
+        """
+        :param IConvergenceAgent agent: Convergence agent to notify of changes.
+        """
+        AMP.__init__(self)
+        self.agent = agent
+
+    def connectionMade(self):
+        self.agent.connected()
+
+    def connectionLost(self, reason):
+        AMP.connectionLost(self, reason)
+        self.agent.disconnected()
```

Comment:
Psh.  And a box receiver, I guess, since this code wants connect/disconnect events.